### ISSUE TITLE
log-cache-cf-auth-proxy crashing frequently due to concurrent map writes

### ISSUE THREAD

**Issue by pianohacker** [cloudfoundry/log-cache-release#28]
On Log Cache 2.1.10, log-cache-cf-auth-proxy is frequently crashing with the following error:

```fatal error: concurrent map writes
fatal error: concurrent map writes
goroutine 177475 [running]:
runtime.throw(0xa674cf, 0x15)
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/runtime/panic.go:774 +0x72 fp=0xc000281600 sp=0xc0002815d0 pc=0x430bd2
runtime.mapassign_faststr(0x996b60, 0xc00008da10, 0xc000305100, 0x645, 0x8)
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/runtime/map_faststr.go:291 +0x3fe fp=0xc000281668 sp=0xc000281600 pc=0x414cde
code.cloudfoundry.org/log-cache/internal/auth.(*UAAClient).Read(0xc00009e460, 0xc000305100, 0x645, 0xc000305100, 0x0, 0x0, 0x0, 0x0)
	/var/vcap/packages/log-cache-cf-auth-proxy/src/code.cloudfoundry.org/log-cache/internal/auth/uaa_client.go:107 +0x911 fp=0xc0002819a0 sp=0xc000281668 pc=0x8fd841
code.cloudfoundry.org/log-cache/internal/auth.CFAuthMiddlewareProvider.Middleware.func1(0xb2c9a0, 0xc0001f40e0, 0xc0001b4300)
	/var/vcap/packages/log-cache-cf-auth-proxy/src/code.cloudfoundry.org/log-cache/internal/auth/cf_auth_middleware.go:88 +0xf0 fp=0xc000281a28 sp=0xc0002819a0 pc=0x8fe380
net/http.HandlerFunc.ServeHTTP(0xc00000f8a0, 0xb2c9a0, 0xc0001f40e0, 0xc0001b4300)
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/net/http/server.go:2007 +0x44 fp=0xc000281a50 sp=0xc000281a28 pc=0x6e8094
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0001d40e0, 0xb2c9a0, 0xc0001f40e0, 0xc0001b4300)
	/var/vcap/packages/log-cache-cf-auth-proxy/src/github.com/gorilla/mux/mux.go:159 +0x104 fp=0xc000281bb8 sp=0xc000281a50 pc=0x86e5c4
code.cloudfoundry.org/log-cache/internal/auth.(*AccessHandler).ServeHTTP(0xc000068680, 0xb2c9a0, 0xc0001f40e0, 0xc000216300)
	/var/vcap/packages/log-cache-cf-auth-proxy/src/code.cloudfoundry.org/log-cache/internal/auth/access_middleware.go:48 +0xf6 fp=0xc000281c18 sp=0xc000281bb8 pc=0x8fa0a6
net/http.serverHandler.ServeHTTP(0xc0001d61c0, 0xb2c9a0, 0xc0001f40e0, 0xc000216300)
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/net/http/server.go:2802 +0xa4 fp=0xc000281c48 sp=0xc000281c18 pc=0x6eb4e4
net/http.(*conn).serve(0xc0002c8320, 0xb2e120, 0xc000104700)
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/net/http/server.go:1890 +0x875 fp=0xc000281fc8 sp=0xc000281c48 pc=0x6e6e85
runtime.goexit()
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000281fd0 sp=0xc000281fc8 pc=0x45faf1
created by net/http.(*Server).Serve
	/var/vcap/data/packages/golang1.13.1/93e1c490dc284ac05cfb026cb706a074fa4a46f3/src/net/http/server.go:2927 +0x38e
```

**Comment by mma-pivotal** [comment]
The original issue was reported by Customer Hitachi Solutions in salesforce ticket 230372.
They said the crash happens 1-2 times every day.

Now I have another customer reporting very similar crash issue from loggr-forwarder-agent.

The ticket number is 231407 from NTT Docomo.

Here is the stack:

```
2020/01/22 05:24:26 Failed to receive data: EOF
fatal error: concurrent map read and map write
goroutine 26 [running]:
runtime.throw(0xa82c93, 0x21)
	/var/vcap/data/packages/golang1.12.4/a751d3f28dc4eb7f861a290aa439356c8b4fca82/src/runtime/panic.go:617 +0x72 fp=0xc00024d6f8 sp=0xc00024d6c8 pc=0x42f0d2
runtime.mapaccess2(0x9b2460, 0xc000115290, 0xc00037de00, 0xc00037de00, 0xc000372700)
	/var/vcap/data/packages/golang1.12.4/a751d3f28dc4eb7f861a290aa439356c8b4fca82/src/runtime/map.go:472 +0x284 fp=0xc00024d740 sp=0xc00024d6f8 pc=0x40f3a4
reflect.mapaccess(0x9b2460, 0xc000115290, 0xc00037de00, 0xa7bdd5)
	/var/vcap/data/packages/golang1.12.4/a751d3f28dc4eb7f861a290aa439356c8b4fca82/src/runtime/map.go:1321 +0x3f fp=0xc00024d778 sp=0xc00024d740 pc=0x4115ff
reflect.Value.MapIndex(0x9b2460, 0xc0002e8500, 0x195, 0x98cac0, 0xc00037de00, 0x98, 0x98cac0, 0xc00037df60, 0x98)
	/var/vcap/data/packages/golang1.12.4/a751d3f28dc4eb7f861a290aa439356c8b4fca82/src/reflect/value.go:1112 +0x10a fp=0xc00024d7d8 sp=0xc00024d778 pc=0x4891ea
github.com/golang/protobuf/proto.makeMapMarshaler.func2(0xc0002e8500, 0x1, 0x0)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:2312 +0x20a fp=0xc00024d8b8 sp=0xc00024d7d8 pc=0x54cdca
github.com/golang/protobuf/proto.(*marshalInfo).size(0xc0002e8380, 0xc0002e84d0, 0xc000075560)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:182 +0xa1 fp=0xc00024d920 sp=0xc00024d8b8 pc=0x523e81
github.com/golang/protobuf/proto.makeMessageSliceMarshaler.func1(0xc00009f140, 0x1, 0xc00024d9a8)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:2239 +0x75 fp=0xc00024d970 sp=0xc00024d920 pc=0x54c5f5
github.com/golang/protobuf/proto.(*marshalInfo).size(0xc0002e82a0, 0xc00009f140, 0xf9b4c0)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:182 +0xa1 fp=0xc00024d9d8 sp=0xc00024d970 pc=0x523e81
github.com/golang/protobuf/proto.(*InternalMessageInfo).Size(0xf9b4c0, 0xb45460, 0xc00009f140, 0x7fc441482648)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:124 +0x67 fp=0xc00024da08 sp=0xc00024d9d8 pc=0x523a27
code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2.(*EnvelopeBatch).XXX_Size(0xc00009f140, 0xb45460)
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2/envelope.pb.go:351 +0x43 fp=0xc00024da38 sp=0xc00024da08 pc=0x88ee93
github.com/golang/protobuf/proto.(*Buffer).Marshal(0xc000358008, 0xb45460, 0xc00009f140, 0xc000396c00, 0xc00009f140)
	/var/vcap/packages/forwarder-agent/src/github.com/golang/protobuf/proto/table_marshal.go:2731 +0x85 fp=0xc00024db10 sp=0xc00024da38 pc=0x537125
google.golang.org/grpc/encoding/proto.marshal(0xa2b500, 0xc00009f140, 0xc000358000, 0x0, 0x0, 0xc000396800, 0x3b5, 0x400)
	/var/vcap/packages/forwarder-agent/src/google.golang.org/grpc/encoding/proto/proto.go:59 +0xe2 fp=0xc00024db58 sp=0xc00024db10 pc=0x85d4a2
google.golang.org/grpc/encoding/proto.codec.Marshal(0xa2b500, 0xc00009f140, 0x400, 0xc00024daa0, 0x3b5, 0xc000075440, 0x0)
	/var/vcap/packages/forwarder-agent/src/google.golang.org/grpc/encoding/proto/proto.go:74 +0xad fp=0xc00024dbd8 sp=0xc00024db58 pc=0x85d5ed
google.golang.org/grpc/encoding/proto.(*codec).Marshal(0xfb8dc8, 0xa2b500, 0xc00009f140, 0xc000032500, 0x0, 0x0, 0xc0000247e0, 0xc00024dca0)
	<autogenerated>:1 +0x46 fp=0xc00024dc20 sp=0xc00024dbd8 pc=0x85de56
google.golang.org/grpc.encode(0x7fc441482600, 0xfb8dc8, 0xa2b500, 0xc00009f140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/var/vcap/packages/forwarder-agent/src/google.golang.org/grpc/rpc_util.go:484 +0x21c fp=0xc00024dcb0 sp=0xc00024dc20 pc=0x8709cc
google.golang.org/grpc.(*csAttempt).sendMsg(0xc000217930, 0xa2b500, 0xc00009f140, 0x0, 0x0)
	/var/vcap/packages/forwarder-agent/src/google.golang.org/grpc/stream.go:457 +0x10d fp=0xc00024ddb8 sp=0xc00024dcb0 pc=0x87c7dd
google.golang.org/grpc.(*clientStream).SendMsg(0xc000104800, 0xa2b500, 0xc00009f140, 0x1, 0xc00009f140)
	/var/vcap/packages/forwarder-agent/src/google.golang.org/grpc/stream.go:379 +0x43 fp=0xc00024ddf0 sp=0xc00024ddb8 pc=0x87c463
code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2.(*ingressBatchSenderClient).Send(0xc0002f0790, 0xc00009f140, 0x0, 0xc000375d88)
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2/ingress.pb.go:201 +0x49 fp=0xc00024de28 sp=0xc00024ddf0 pc=0x893179
code.cloudfoundry.org/go-loggregator.(*IngressClient).emit(0xc0000b1360, 0xc00009a520, 0x1, 0x1, 0x0, 0x22684f4c1d3200)
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/ingress_client.go:551 +0x89 fp=0xc00024de88 sp=0xc00024de28 pc=0x89cb09
code.cloudfoundry.org/go-loggregator.(*IngressClient).flush(0xc0000b1360, 0xc00009a520, 0x1, 0x1, 0x1, 0xc00009a520)
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/ingress_client.go:534 +0x4d fp=0xc00024dee0 sp=0xc00024de88 pc=0x89c98d
code.cloudfoundry.org/go-loggregator.(*IngressClient).startSender(0xc0000b1360)
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/ingress_client.go:518 +0x42c fp=0xc00024dfd8 sp=0xc00024dee0 pc=0x89c8ec
runtime.goexit()
	/var/vcap/data/packages/golang1.12.4/a751d3f28dc4eb7f861a290aa439356c8b4fca82/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc00024dfe0 sp=0xc00024dfd8 pc=0x45d761
created by code.cloudfoundry.org/go-loggregator.NewIngressClient
	/var/vcap/packages/forwarder-agent/src/code.cloudfoundry.org/go-loggregator/ingress_client.go:145 +0x3c6
```

For NTT Docomo, this is a one time issue and they are using PAS 2.6.4, which uses log-cahce 2.1.6.

**Comment by pianohacker** [comment]
For context, @mma-pivotal 's loggr-forwarder-agent crash is https://github.com/cloudfoundry/loggregator-agent-release/issues/26 .

This issue should be fixed in https://github.com/cloudfoundry/log-cache-release/commit/c3339a997eeceb14ffc1cb3dd2080010b74f8137 , released in log-cache v2.1.13.
